README.md
<wiki:toc max_depth="4" />
#The TMVS program user guide

= Introduction =

    # Input / Output file format
    # Config parameters
    # Command line instruction


= Details =

==Input / Output file format==

===Input file format===

Our input file format includes "nvm", "nvm2" and "mvs"

====nvm file format====
NVM file is come from [http://ccwu.me/vsfm/ VisualSFM] data structure.
Here is the example of the data structure. The detail can be found in [https://code.google.com/p/pais-mvs/source/browse/trunk/TMVS/io/fileloader.cpp FileLoader].

You can skip the calibration step and just use the Structure from Motion result output from [http://ccwu.me/vsfm/ VisualSFM].

{{{
NVM_V3 

5
pawn0013.jpg	614.095397949 0.705410371683 0.160690743319 0.671401589359 0.160605237544 -0.556085150075 0.0481223921551 -0.00781510757143 -0.199289312888 0 
pawn0010.jpg	616.175537109 0.90353903514 0.221746421078 0.3576944596 0.0806247263945 -0.880841878288 0.0327703491031 -0.684201024844 -0.209314043486 0 
pawn0011.jpg	612.03302002 0.85241383667 0.2037593266 0.469072019941 0.108830220502 -0.71971232163 0.0433857776889 -0.492035476323 -0.207263977174 0 
pawn0012.jpg	611.360473633 0.786507583571 0.183363764635 0.573952646995 0.135504187104 -0.608685012281 0.0487066227347 -0.263440114899 -0.203210786458 0 
pawn0014.jpg	617.585876465 0.611485687162 0.135944898976 0.757586998462 0.183482834469 -0.572254659063 0.0434025057556 0.255716172724 -0.198563271584 0 
}}}

The first row shows the NVM file version.

Third line "5" indicate there are 5 cameras in the following list.

Each list entry:

 # image file name, can be PNG, JPG, BMP, or any OpenCV supported image format. Note that the file path can be relative or absolute path. 
 # focal length
 # 4 elements vector (k wx wy wz) indicate the quaternion rotation. Transformation form quaternion to 3x3 rotation matrix can be found in code. [https://code.google.com/p/pais-mvs/source/browse/trunk/TMVS/mvs/camera.cpp#6 Camera].
 # 3 elements vector (x y z) indicate camera center in world coordinate.
 # radial distortion parameter. The usage of parameter can be found in [https://code.google.com/p/pais-mvs/source/browse/trunk/TMVS/mvs/camera.cpp#138 camera projection].
 # 0 for end of line


====nvm2 file format====
In the NVM file format, the camera principle point is assumed at the image center, and can't to be edit.

NVM2 file specify the principle point(px, py) after focal length. Others the same.
 
====mvs file format====
MVS file is saved in a binary way for internal use, for example: MVS viewer, patch filtering.

===Output file format===

Our output file format includes "mvs", "ply" and "psr"

====mvs file format====
MVS file is saved in a binary way for internal use, for example: MVS viewer, patch filtering.

====ply file format====
ply is the point cloud model. Can be open using MeshLab.

====psr file format====
psr file is ready for the [http://www.cs.jhu.edu/~misha/Code/PoissonRecon/Version5.71/ Poisson surface reconstruction].

==Config parameters==
config.txt should be placed in the same directory of main program.

{{{
### patch optimization configuration ###
# patch radius (patch size = 2*radius+1)
# !!!!check distWeighting if patchRadius have been changed!!!!
# default 15
patchRadius			15

# reduce normal search range for expansion patch (reduce to +- pi/reduceNormalRange)
# 1 stands for search 360 degree, 2 for 180 degree
# smaller for larger cell size
# default 2 (reduce to 180 degree)
reduceNormalRange	2

# enable or disable adaptive distance weighting (0: disable, 1: enable)
# default 1
adaptiveDistanceEnable		1
# distance weighting to patch center
# default(patch radius/3) 5
distWeighting		5

# enable or disable adaptive difference weighting (0: disable, 1: enable)
# default 1
adaptiveDifferenceEnable	1
# difference weighting
# default(128*128) 16384
diffWeighting		16384

# enable or disable adaptive gradient weighting (0: disable, 1: enable)
# default 0
adaptiveGradientEnable		0
# gradient maginitude weighting
# default 10.0
gradientWeighting	10.0

# visible camera angle correlation
# smaller for sparse camera deploy
# default 0.87 radian for 30 degree
visibleCorrelation	0.7

# depth search range scalar
# smaller for smoother object, larger for complex/sharp object
# smaller for small cell size, larger for big cell size
# larger depthRangeScalar cloud avoid local trap!
# beta in paper
# default(cellsize/2) 2
depthRangeScalar	8

### PSO optimization configuration ###
# particle number
# default 30
particleNum			15
# maximum iteration number
# default 60
maxIteration		30

### expansion configuration ###
# image cell size, control the density of patches
# default 4
cellSize			2
# maximum patch number in single cell
# default 3
maxCellPatchNum		3
# patch expansion strategy (Best: 0, Worst: 1, Breath: 2, Depth: 3)
# default 0
expansionStrategy	0

### level of detail configuration ###
# texture variation
# default 36
textureVariation	36
# minimum level of detail
# default 0
minLOD				0
# maximum level of detail
# default 15
maxLOD				15
# level of detail ratio
# default 0.8
lodRatio			0.8

### filtering configuration ###
# minimum visible camera number
# default 3
minCamNum			3
# minimum texture correlation
# default 0.9
minCorrelation		0.9
# minimum patch homography ellipse axes ratio
# default 0.15
minRegionRatio		0.15
# maximum fitness (critical effect the completeness)
# default 10.0
maxFitness			10.0
# neighbor patch search radius scalar (from TVCG09 PCMVS)
# the neighbor radius = pow(volume, 1.0/3.0) * neighborRadiusScalar
# default 0.0025*(cellsize*2)
neighborRadiusScalar		0.01
}}}


TMVS/ChangeLog.txt
2012/08/09
* config loader
* linear patch priority
* expansion strategy
* getPatch() return pointer 
* auto normal length in mvsviewer
* depth search range narrow view projection threshold 1 -> 0.01
2012/06/13
* BUG: loader correlation table -> homography -> warping overflow (fixed)
2012/06/12
* BUG: expansion -> correlation table -> maxIdx (fitness = 0) (fixed)
2012/06/01
* Fix boot bug (static variable)
2012/05/31
* Add remove invisible camera.
* Add toggle viewer normal.
* Add point pick function.
2012/05/30
* Add remove invisible camera check.
2012/05/28
* Add adapted weighting
2012/05/27
* Add SVN project
2012/05/24
* Add MVS viewer and loader
2012/05/19
* Add patch refinement with LOD
2012/05/15
* Add PSO solver
2012/05/11
* To Do:
    1. add fitness function epipolar constrain
    2. add fitness function color weighting
2012/05/05
* prepared for rebuild project.
2012/05/02
* viewer color toggle
2012/05/01
* add Viewer point color.
* add single seed expansion.
2012/04/30
* expansion with visible cameras.
* Add viewer pick point function.
* expansion filtering with visible cameras normal.
2012/04/28
* fix priority queue bug.
2012/04/27
* add feature to write PLY file
* add point color.
2012/04/24
* Update cell projection + priority (add visibility test to all seeds)
2012/04/19
* Add an experiment version of depth test
* re-activate neighbor cell fail loops
2012/04/18
* remove patch sampling information saved
2012/04/16
* To Do:
    1. Gaussian init for PSO
    2. Z-Test
2012/04/13
* Add a CPU register optimized warping function
* Fix PSO init particle bug.
2012/04/12
* A PSO fitness CUDA version (Fail&fixed homography not test)
2012/04/11 - 2
* CUDA avaliable with 2D layered texture!
2012/04/11
* a stable version
2012/04/10 - 2
* Seed refinement works
2012/04/10
* change PSO to 0326 version (Not work), changed back
* Add CUDA texture function.
* Add CUDA warping test function.
2012/04/09
* Integrate warping function of seed fitness function and expansion fitness function
2012/04/07
* Seed refinement only normal.
* Tried Huber loss function in seed fitness, but not work. (0326 version works)
2012/04/05
* Enhance viewer functions.
2012/03/28
* A experiment version of expansion
2012/03/27
* Giveup CUDA 4.1 Orz
* 3D vector quantization and filtering in parameter space
2012/03/24
* Move to 64bit platform
* Integrate CUDA 4.1
2012/03/22
* To Do:
    1. sparse point initialization and refinement
    2. seed point filtering
    3. point expansion
2012/03/21
* add cell maps
2012/03/17
* patch filtering in parameter space
2012/03/16
* mvs file read/write
2012/03/14
* add depth adjustment range
* GLN-PSO add depth dimension
* fix gaussian weighting (normalize)
2012/03/13
* point filter (re-projection error and epipolar line distance)
2012/03/12
* Add full search in GLN-PSO
* Advanced GLN-PSO visualization
* Fix GLN-PSO localK bug
2012/03/10
* Fundamental matrix
2012/03/08
* GLN-PSO bug fixed
2012/03/07
* GLN-PSO still buggy
2012/03/06
* add background mask to std filtering
* performance fix for filtering
2012/03/05
* start editing expansion and filtering
2012/03/04
* Threshold std image
2012/03/03
* Faster image warping
2012/03/02
* intersetion full pixel-wised std (Geometric Bound)
2012/03/01
* full image pixel-wised std
2012/02/29
* PSO performance enhancement
2012/02/25
* PCL Visualizer (press "H" for help)
2012/02/24
* Fix random number gereration problem.
* Project 3D plane point.
* To Do:
  1. Reduce normal range (phi theda)
  2. PSO memory optimization (Local Best & Near Best)
  3. openMP sync problem in GLN-PSO (need check!)
  4. PSO parameter space and particle movement visualization
  5. Faster reference camera desicion

TMVS/TMVS.cpp
#include "stdafx.h"

#include <time.h>

#include <opencv2\opencv.hpp>

#include "mvs\mvs.h"
#include "view\mvsviewer.h"
#include "mvs\featuremanager.h"

using namespace cv;
using namespace PAIS;

ofstream debugFile("debug.txt", ofstream::out);

MvsViewer *viewer;

void addPatchView(const Patch &pth) {
	if (viewer != NULL) {
		viewer->addPatch(pth);
	}
}

int main(int argc, char* argv[])
{
	// MVS configures
	MvsConfig config;
	config.cellSize           = 4;
	config.patchRadius        = 15;
	config.reduceNormalRange  = 2;
	config.adaptiveEnable     = true;
	config.distWeighting      = config.patchRadius / 3.0;
	config.diffWeighting      = 128*128;
	config.minCamNum          = 3;
	config.textureVariation   = 36;
	config.visibleCorrelation = 0.7;
	config.minCorrelation     = 0.7;
	config.maxFitness         = 10.0;
	config.minLOD             = 0;
	config.maxLOD             = 15;
	config.lodRatio           = 0.8;
	config.maxCellPatchNum    = 3;
	config.neighborRadius     = 0.005;
	config.minRegionRatio     = 0.55;
	config.depthRangeScalar   = 1;
	config.particleNum        = 5;
	config.maxIteration       = 10;
	config.expansionStrategy  = MVS::EXPANSION_BEST_FIRST;

	FileLoader::loadConfig("config.txt", config);

	// set MVS instance
	MVS &mvs = MVS::getInstance(config);

	//mvs.loadNVM2("D:/workspace/TMVS_data/boxball/boxball.nvm2");
	//FeatureManager::getFeatureDescriptor(mvs.getCameras());

	if (argc >= 3) {
		if ( strcmp(argv[1], "-v") == 0 ) {         // viewer
			mvs.loadMVS((char*)argv[2]);
			viewer = new MvsViewer(mvs, true, true, false);
			viewer->open();

		} else if ( strcmp(argv[1], "-a") == 0 ) {  // animate
			mvs.loadMVS((char*)argv[2]);
			viewer = new MvsViewer(mvs, true, true, true);
			viewer->open();

		} else if ( strcmp(argv[1], "-r") == 0 ) {  // reconstruction
			string fileName(argv[2]);
			size_t found = fileName.find_last_of(".");
			string fileExt = fileName.substr(found+1);

			// load file
			if ( fileExt.compare("nvm") == 0 ) {
				mvs.loadNVM(argv[2]);
			} else if ( fileExt.compare("nvm2") == 0 ) {
				mvs.loadNVM2(argv[2]);
			} else if ( fileExt.compare("mvs") == 0 ) {
				mvs.loadMVS(argv[2]);
			}

			printf("patches: %d\n", mvs.getPatches().size());

			// run reconstruction
			clock_t start_t, end_t;
			start_t = clock();
			mvs.writeMVS("init.mvs");
			mvs.refineSeedPatches();
			mvs.writeMVS("seed.mvs");
			mvs.expansionPatches();
			mvs.writeMVS("exp.mvs");
			mvs.writePLY("exp.ply");
			mvs.writePSR("exp.psr");
			/*
			mvs.cellFiltering();
			printf("patches: %d\n", mvs.getPatches().size());
			mvs.visibilityFiltering();
			printf("patches: %d\n", mvs.getPatches().size());
			mvs.neighborCellFiltering(0.25);
			printf("patches: %d\n", mvs.getPatches().size());
			mvs.writeMVS("filter.mvs");
			mvs.neighborPatchFiltering();
			mvs.writeMVS("filter2.mvs");
			mvs.writePLY("cloud.ply");
			mvs.writePSR("cloud.psr");
			*/
			end_t = clock();
			double totime = (double)(end_t - start_t) / CLOCKS_PER_SEC;
			printf("time1\t%f\n", totime);
			debugFile << "total time: " << totime << endl;
			system("pause");
		}
	} else {
		// Todo: useage message
		char *msg = "-v [filename.mvs]: viewer\n-a [filename.mvs]: animate\n-r {[filename.mvs], [filename.nvm], [filename.nvm2]}: reconstruction\n";
		printf(msg);
	}

	debugFile.close();
	return 0;
}